File: sdks/typescript/src/resources/memories.ts
/** Memory CRUD + audit (POST/GET/PATCH/DELETE /v1/memories, .../audit). */

import type { HttpClient } from "../http.js";
import type {
  AuditLog,
  CreateMemoryParams,
  ListMemoriesParams,
  Memory,
  MemoryPage,
  UpdateMemoryParams,
} from "../types.js";

export class MemoriesResource {
  constructor(private readonly http: HttpClient) {}

  create(params: CreateMemoryParams): Promise<Memory> {
    const { namespace, content, type = "fact", metadata, source } = params;
    const body: Record<string, unknown> = { content, namespace, type };
    if (metadata !== undefined) body.metadata = metadata;
    if (source !== undefined) body.source = source;
    return this.http.request<Memory>("POST", "/v1/memories", { body });
  }

  get(memoryId: string, namespace?: string): Promise<Memory> {
    return this.http.request<Memory>("GET", `/v1/memories/${encodeURIComponent(memoryId)}`, {
      params: { namespace },
    });
  }

  update(memoryId: string, params: UpdateMemoryParams): Promise<Memory> {
    const { namespace, ...rest } = params;
    const body: Record<string, unknown> = {};
    if (rest.content !== undefined) body.content = rest.content;
    if (rest.type !== undefined) body.type = rest.type;
    if (rest.metadata !== undefined) body.metadata = rest.metadata;
    return this.http.request<Memory>("PATCH", `/v1/memories/${encodeURIComponent(memoryId)}`, {
      params: { namespace },
      body,
    });
  }

  async delete(memoryId: string, opts: { hard?: boolean; namespace?: string } = {}): Promise<void> {
    await this.http.request<void>("DELETE", `/v1/memories/${encodeURIComponent(memoryId)}`, {
      params: { hard: opts.hard ?? false, namespace: opts.namespace },
    });
  }

  list(params: ListMemoriesParams): Promise<MemoryPage> {
    const { namespace, type, state, limit = 50, offset = 0 } = params;
    return this.http.request<MemoryPage>("GET", "/v1/memories", {
      params: { namespace, type, state, limit, offset },
    });
  }

  audit(memoryId: string): Promise<AuditLog> {
    return this.http.request<AuditLog>(
      "GET",
      `/v1/memories/${encodeURIComponent(memoryId)}/audit`,
    );
  }
}

File: sdks/python/src/scp_memory_sdk/resources/memories.py
"""Memory CRUD + audit (POST/GET/PATCH/DELETE /v1/memories, .../audit)."""

from __future__ import annotations

from scp_memory_sdk._http import HttpClient
from scp_memory_sdk.models import AuditLog, Memory, MemoryPage


class MemoriesResource:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def create(
        self,
        *,
        content: str,
        namespace: str,
        type: str = "fact",
        metadata: dict | None = None,
        source: str | None = None,
    ) -> Memory:
        """Create a memory (emits an audit event server-side)."""
        body: dict = {"content": content, "namespace": namespace, "type": type}
        if metadata is not None:
            body["metadata"] = metadata
        if source is not None:
            body["source"] = source
        return Memory.from_dict(self._http.request("POST", "/v1/memories", json=body))

    def get(self, memory_id: str, *, namespace: str | None = None) -> Memory:
        return Memory.from_dict(
            self._http.request("GET", f"/v1/memories/{memory_id}", params={"namespace": namespace})
        )

    def update(
        self,
        memory_id: str,
        *,
        content: str | None = None,
        type: str | None = None,
        metadata: dict | None = None,
        namespace: str | None = None,
    ) -> Memory:
        body: dict = {}
        if content is not None:
            body["content"] = content
        if type is not None:
            body["type"] = type
        if metadata is not None:
            body["metadata"] = metadata
        return Memory.from_dict(
            self._http.request(
                "PATCH",
                f"/v1/memories/{memory_id}",
                params={"namespace": namespace},
                json=body,
            )
        )

    def delete(self, memory_id: str, *, hard: bool = False, namespace: str | None = None) -> None:
        self._http.request(
            "DELETE",
            f"/v1/memories/{memory_id}",
            params={"hard": hard, "namespace": namespace},
        )

    def list(
        self,
        *,
        namespace: str,
        type: str | None = None,
        state: str | None = None,
        limit: int = 50,
        offset: int = 0,
    ) -> MemoryPage:
        return MemoryPage.from_dict(
            self._http.request(
                "GET",
                "/v1/memories",
                params={
                    "namespace": namespace,
                    "type": type,
                    "state": state,
                    "limit": limit,
                    "offset": offset,
                },
            )
        )

    def audit(self, memory_id: str) -> AuditLog:
        return AuditLog.from_dict(self._http.request("GET", f"/v1/memories/{memory_id}/audit"))
